### PR TITLE
Add missing members to Data.Array.NonEmpty

### PR DESCRIPTION
**Pull request by kl0tl**
We added a few functions to Data.Array in https://github.com/purescript/purescript-arrays/pull/189, https://github.com/purescript/purescript-arrays/pull/188 and https://github.com/purescript/purescript-arrays/pull/68 but forgot to export variants from Data.Array.NonEmpty.

This pull request adapts them to non empty arrays and also re-exports foldl1, foldr1, foldMap1, fold1 and intercalate (from Data.Semigroup.Foldable) so that Data.Array and Data.Array.NonEmpty have (mostly) the same interface.

### CODE REVIEW

**Comment by hdgarrood** [comment]
I think we should export monomorphic versions of the fold* functions in both this module and in Data.Array, see https://github.com/purescript/purescript-arrays/pull/173.

**Comment by JordanMartinez** [comment]
Data.Array.NonEmpty does not have `groupAll` or `groupAllBy`. I'm assuming we want these included there, too, correct?

**Comment by JordanMartinez** [comment]
I've updated `Data.Array` and `Data.Array.NonEmpty` to export monomorphic versions of `foldl`, `foldr`, `fold1`, `foldMap`, and `intercalate`.

**Comment by JordanMartinez** [comment]
> Data.Array.NonEmpty does not have groupAll or groupAllBy. I'm assuming we want these included there, too, correct?

Nevermind. This specific PR doesn't include those members in its exports because it was based on a previous branch. Once I merged it with `master`, that was no longer the case.

File: test/Test/Data/Array/NonEmpty.purs [hunk 1]
```diff
@@ -106,6 +114,15 @@ testNonEmptyArray = do
   assert $ NEA.elemLastIndex 1 (fromArray [1, 2, 1]) == Just 2
   assert $ NEA.elemLastIndex 4 (fromArray [1, 2, 1]) == Nothing
 
+  log "find should return the first element for which a predicate returns true in an array"
+  assert $ NEA.find (_ /= 1) (fromArray [1, 2, 1]) == Just 2
+  assert $ NEA.find (_ == 3) (fromArray [1, 2, 1]) == Nothing
```
Comment (JordanMartinez) [inline test/Test/Data/Array/NonEmpty.purs:123]: Should there be another test here to verify that the first element is returned when two or more elements would satisfy the predicate?

**Review by JordanMartinez — APPROVED** [review]

**Comment by kl0tl** [comment]
Shouldn’t we rather remove the re-exports? That’s quite inconsistent with https://github.com/purescript/purescript-nonempty/pull/42, where we deprecated the monomorphic Data.NonEmpty.foldl1, otherwise.

File: test/Test/Data/Array/NonEmpty.purs (hunk 1 above)
Comment (kl0tl) [reply test/Test/Data/Array/NonEmpty.purs:123]: I think so. We could also ensure that findIndex actually returns the index of the first match by changing the `_ /= 1` predicate to `_ == 1`  .

**Comment by hdgarrood** [comment]
I suppose it is but if we want these to be consistent then we should reintroduce the NonEmpty ones. Removing these re-exports from Data.Array is much too breaking to be worthy of consideration in my mind.

**Comment by JordanMartinez** [comment]
> Shouldn’t we rather remove the re-exports?

I think that was deprecated because, while the `NonEmpty` part is monomorphic, its underlying `f` is still polymorphic. Here, we have a monomorphic `Array` or `NonEmptyArray`.

File: test/Test/Data/Array/NonEmpty.purs (hunk 1 above)
Comment (JordanMartinez) [reply test/Test/Data/Array/NonEmpty.purs:123]: Fixed

**Review by hdgarrood — APPROVED** [review]

**Comment by JordanMartinez** [comment]
I'll open an PR that re-adds the monomorphic `fold1` to NonEmpty

**Comment by JordanMartinez** [comment]
See https://github.com/purescript/purescript-nonempty/pull/45

**Pull request merged**